frontend/src/metabase-lib/lib/Dimension.ts

```diff
@@ -815,80 +818,127 @@ export class FieldDimension extends Dimension {
     return field;
   }
 
-  field(): Field {
+  _getIdentifierProp() {
+    return this.isIntegerFieldId() ? "id" : "name";
+  }
+
+  _getTrustedFieldCachedOnInstance() {
     if (
       this._fieldInstance &&
       this._fieldInstance._comesFromEndpoint === true
     ) {
       return this._fieldInstance;
     }
+  }
 
-    const question = this.query()?.question();
-    const lookupField = this.isIntegerFieldId() ? "id" : "name";
-    const fieldMetadata = question
-      ? _.findWhere(question.getResultMetadata(), {
-          [lookupField]: this.fieldIdOrName(),
-        })
-      : null;
+  _getFieldMetadataFromSavedDataset() {
+    const identifierProp = this._getIdentifierProp();
+    const questionAssociatedWithDimension = this.query()?.question();
+    if (
+      questionAssociatedWithDimension?.isSaved() &&
+      questionAssociatedWithDimension.isDataset()
+    ) {
+      const question = this.query()?.question();
+      const field = _.findWhere(question.getResultMetadata(), {
+        [identifierProp]: this.fieldIdOrName(),
+      });
 
-    // Field result metadata can be overwritten for models,
-    // so we need to merge regular field object with the model overwrites
-    const shouldMergeFieldResultMetadata = question?.isDataset();
+      return field;
+    }
+  }
 
-    if (this.isIntegerFieldId()) {
-      const field = this._metadata?.field(this.fieldIdOrName());
-
-      if (field) {
-        if (!fieldMetadata || !shouldMergeFieldResultMetadata) {
-          return field;
-        }
-        const fieldObject = merge(
-          field instanceof Field ? field.getPlainObject() : field,
-          fieldMetadata,
-        );
-        return this._createField(fieldObject, { hydrate: true });
-      }
+  _getFieldMetadataFromNestedDataset() {
+    const identifierProp = this._getIdentifierProp();
+    const virtualTableCardId = getQuestionIdFromVirtualTableId(
+      this.query()?.sourceTableId?.(),
+    );
+    if (virtualTableCardId != null) {
+      const question = this._metadata?.question(virtualTableCardId);
+      const field =
+        question && question.isDataset()
```

Comment:
should we use optional chaining like the rest of the change?

Reply:
Updated